common/src/main/java/org/openmrs/analytics/FhirStoreUtil.java

```diff
@@ -46,118 +34,92 @@
+	private final String sinkUsername;
 	
-	private static final GsonFactory JSON_FACTORY = new GsonFactory();
+	private final String sinkPassword;
+	
+	private static final Pattern GCP_PATTERN = Pattern
+	        .compile("projects/[\\w-]+/locations/[\\w-]+/datasets/[\\w-]+/fhirStores/[\\w-]+");
 	
-	private static final NetHttpTransport HTTP_TRANSPORT = new NetHttpTransport();
+	protected FhirStoreUtil(String sinkUrl, String sinkUsername, String sinkPassword, IRestfulClientFactory clientFactory) {
+		this.clientFactory = clientFactory;
+		this.sinkUrl = sinkUrl;
+		this.sinkUsername = sinkUsername;
+		this.sinkPassword = sinkPassword;
+	}
 	
-	private HttpUtil httpUtil;
+	public static FhirStoreUtil createFhirStoreUtil(String sinkUrl, IRestfulClientFactory clientFactory)
+	        throws IllegalArgumentException {
+		return createFhirStoreUtil(sinkUrl, "", "", clientFactory);
+	}
 	
-	FhirStoreUtil(String gcpFhirStore, FhirContext fhirContext) throws IllegalArgumentException {
-		// TODO separate the GCP specific functionality out of this class.
-		this.fhirContext = fhirContext;
-		
-		Matcher fhirMatcher = FHIR_PATTERN.matcher(gcpFhirStore);
-		if (!fhirMatcher.matches()) {
-			throw new IllegalArgumentException(
-			        String.format("The gcpFhirStore %s does not match %s pattern!", gcpFhirStore, FHIR_PATTERN));
+	public static FhirStoreUtil createFhirStoreUtil(String sinkUrl, String sinkUsername, String sinkPassword,
+	        IRestfulClientFactory clientFactory) throws IllegalArgumentException {
+		if (matchesGcpPattern(sinkUrl)) {
+			return new GcpStoreUtil(sinkUrl, clientFactory);
+		} else {
+			return new FhirStoreUtil(sinkUrl, sinkUsername, sinkPassword, clientFactory);
 		}
-		
-		this.gcpFhirStore = gcpFhirStore;
-		this.httpUtil = new HttpUtil();
 	}
 	
-	// This follows the examples at:
-	// https://github.com/GoogleCloudPlatform/java-docs-samples/healthcare/tree/master/healthcare/v1
-	// TODO: remove redundant resource information if passing a HAPI resource
-	public void uploadResourceToCloud(String resourceType, String resourceId, Resource resource) {
-		try {
-			updateFhirResource(gcpFhirStore, resourceId, resourceType, resource);
+	public MethodOutcome uploadResource(Resource resource) {
+		Collection<IClientInterceptor> interceptors = Collections.<IClientInterceptor> emptyList();
+
+		if (sinkUsername != null && !sinkUsername.isEmpty() && sinkPassword != null && !sinkPassword.isEmpty()) {
+			interceptors = Collections.<IClientInterceptor> singleton(new BasicAuthInterceptor(sinkUsername, sinkPassword));
 		}
-		catch (IOException e) {
-			log.error(String.format("IOException while using Google APIs: %s", e.toString()));
+		
+		try {
+			return updateFhirResource(sinkUrl, resource, interceptors);
 		}
-		catch (URISyntaxException e) {
-			log.error(String.format("URI syntax exception while using Google APIs: %s", e.toString()));
+		catch (Exception e) {
```

Comment:
Are we expecting any exceptions here? Or put it another way: Are you trying to catch all runtime exceptions here? If yes, why? I think it is a better practice to only catch checked exceptions.